drivers/memc/memc_nxp_flexram.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/devicetree.h>
+#include <soc.h>
+
+#define FLEXRAM_DT_NODE DT_INST(0, nxp_flexram)
+#define IOMUXC_GPR_DT_NODE DT_INST(0, nxp_imx_gpr)
+
+#if defined(CONFIG_MEMC_NXP_FLEXRAM_MAGIC_ADDR_API) || \
+	defined(CONFIG_MEMC_NXP_FLEXRAM_ERROR_INTERRUPT)
+#define FLEXRAM_INTERRUPTS_USED
+#endif
+
+#if DT_PROP_HAS_IDX(FLEXRAM_DT_NODE, flexram_bank_spec, 0)
+#define FLEXRAM_RUNTIME_BANKS_USED 1
+#endif
+
+#ifdef FLEXRAM_INTERRUPTS_USED
+enum memc_flexram_interrupt_cause {
+#ifdef CONFIG_MEMC_NXP_FLEXRAM_ERROR_INTERRUPT
+	flexram_ocram_access_error,
+	flexram_itcm_access_error,
+	flexram_dtcm_access_error,
+#endif
+#ifdef CONFIG_MEMC_NXP_FLEXRAM_MAGIC_ADDR_API
+	flexram_ocram_magic_addr,
+	flexram_itcm_magic_addr,
+	flexram_dtcm_magic_addr,
+#endif /* CONFIG_MEMC_NXP_FLEXRAM_MAGIC_ADDR_API */
+};
+
+typedef void (*flexram_callback_t)(enum memc_flexram_interrupt_cause, void *user_data);
+
+void memc_flexram_register_callback(flexram_callback_t callback, void *user_data);
+#endif /* FLEXRAM_INTERRUPTS_USED */
+
+#ifdef FLEXRAM_RUNTIME_BANKS_USED
+
+/*
+ * call from platform_init to set up flexram if using runtime map
+ * must be inlined because cannot use stack
+ */
+#define GPR17_REG_FILL(node_id, prop, idx) + (DT_PROP_BY_IDX(node_id, prop, idx) << idx)
+static inline void memc_flexram_dt_partition(void)
+{
+	/* iomuxc_gpr must be const (in ROM region) because used in reconfiguring ram */
+	static IOMUXC_GPR_Type *const iomuxc_gpr =
+		(IOMUXC_GPR_Type *) DT_REG_ADDR(IOMUXC_GPR_DT_NODE);
+	/* do not create stack variables or use any data from ram in this function */
+	iomuxc_gpr->GPR17 = DT_FOREACH_PROP_ELEM_SEP(FLEXRAM_DT_NODE,
+				flexram_bank_spec, GPR17_REG_FILL, (+));
+	iomuxc_gpr->GPR16 |= IOMUXC_GPR_GPR16_FLEXRAM_BANK_CFG_SEL_MASK;
+}
+#endif /* FLEXRAM_RUNTIME_BANKS_USED */
+
+
+#ifdef CONFIG_MEMC_NXP_FLEXRAM_MAGIC_ADDR_API
+/* 0 on success, negative on error */
+int memc_flexram_set_ocram_magic_addr(uint16_t ocram_addr);
```

Comment:
Please add more documentation about what the "Magic address" does- from the RT1170 reference manual, it seems it will generate an interrupt on an read or a write:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/abd8f7c6-0e56-4a8c-9752-60e9201c76bb)

The API also seems to only support interrupting on a read currently, but the hardware supports either mode

Reply:
The sample shows it working for read and write.

Reply:
Understood- if the hardware generates an interrupt on read and write, then I'm good with the current API. The reference manual seems to imply that the value of `OCRAM_WR_RD_SEL` would determine whether the interrupt occurs on a read or on a write

Reply:
to be honest, it looks like that bit is affecting nothing, but, I will check the assembly to make sure there aren't two reads for some reason, but that would be a bad compilation of the sample imo

Reply:
yes, the assembly clearly causes the interrupt from a strb instruction, which is a write, so, I don't know what the story of this bit is